internal/freelist/freelist.go

```diff
@@ -0,0 +1,100 @@
+package freelist
+
+import (
+	"sort"
+
+	"go.etcd.io/bbolt/internal/common"
+)
+
+type ReadWriter interface {
+	// Read calls Init with the page ids stored in te given page.
+	Read(page *common.Page)
+
+	// Write writes the freelist into the given page.
+	Write(page *common.Page)
+
+	// EstimatedWritePageSize returns the size of the freelist after serialization in Write.
+	// This should never underestimate the size.
+	EstimatedWritePageSize() int
+}
+
+type Interface interface {
+	ReadWriter
+
+	// Init initializes this freelist with the given list of pages.
+	Init(ids common.Pgids)
+
+	// Allocate returns the starting page id of a contiguous block of a given size in number of pages.
+	// If a contiguous block cannot be found then 0 is returned.
+	Allocate(txid common.Txid, numPages int) common.Pgid
+
+	// Count returns the number of free and pending pages.
+	Count() int
+
+	// FreeCount returns the number of free pages.
+	FreeCount() int
+
+	// FreePageIds returns all free pages.
+	FreePageIds() common.Pgids
+
+	// PendingCount returns the number of pending pages.
+	PendingCount() int
+
+	// Release moves all page ids for a transaction id (or older) to the freelist.
+	Release(txId common.Txid)
+
+	// ReleaseRange moves pending pages allocated within an extent [begin,end] to the free list.
+	ReleaseRange(begin, end common.Txid)
```

Comment:
Once https://github.com/etcd-io/bbolt/pull/783 gets merged, then no need to export such interface methods.

Reply:
put it into shared, but keeping it package private for the unit tests that still depend on this